mountpoint-s3-crt/src/checksums/crc32.rs

```diff
@@ -0,0 +1,76 @@
+use mountpoint_s3_crt_sys::*;
+
+/// CRC32 checksum
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Crc32(u32);
+
+/// Computes the CRC32 checksum of a byte slice.
+///
+/// Use [`Hasher`] for more advanced use-cases.
+pub fn hash(buf: &[u8]) -> Crc32 {
+    let mut hasher = Hasher::new();
+    hasher.update(buf);
+    hasher.finalize()
+}
+
+/// CRC32 Hasher
+#[derive(Debug, Clone)]
+pub struct Hasher {
+    state: Crc32,
+}
+
+impl Hasher {
+    /// Create a new CRC32 [`Hasher`].
+    pub fn new() -> Self {
+        Self { state: Crc32(0) }
+    }
+
+    /// Update the hash state with the given bytes slice.
+    pub fn update(&mut self, buf: &[u8]) {
+        self.state = Hasher::crc32(buf, &self.state);
+    }
+
+    /// Finalize the hash state and return the computed CRC32 checksum value.
+    pub fn finalize(self) -> Crc32 {
+        self.state
+    }
+
+    /// Compute CRC32 checksum of the data in the given bytes slice, append to the previous checksum.
+    fn crc32(buf: &[u8], previous_checksum: &Crc32) -> Crc32 {
```

Comment:
We need to `assert!(buf.len() <= i32::MAX);` here to make the `buf.len() as i32` cast safe, and probably also mention that bound in the comment.